lib/Backend/Lower.cpp

```diff
@@ -23001,6 +22972,242 @@ Lowerer::LoadIndexFromLikelyFloat(
+    InsertTestBranch(objectArrayOpnd, IR::IntConstOpnd::New((uint32)Js::DynamicObjectFlags::ObjectArrayFlagsTag, TyUint8, func), 
+        Js::OpCode::BrNeq_A, noObjectArrayLabel, insertBeforeInstr);
+
+    IR::RegOpnd * typeHandlerOpnd = IR::RegOpnd::New(TyMachPtr, func);
+    InsertMove(typeHandlerOpnd, IR::IndirOpnd::New(typeOpnd, Js::DynamicType::GetOffsetOfTypeHandler(), TyMachPtr, func), insertBeforeInstr);
+    InsertCompareBranch(IR::IndirOpnd::New(typeHandlerOpnd, Js::DynamicTypeHandler::GetOffsetOfOffsetOfInlineSlots(), TyUint16, func),
+        IR::IntConstOpnd::New(Js::DynamicTypeHandler::GetOffsetOfObjectHeaderInlineSlots(), TyUint16, func), 
+        Js::OpCode::BrNeq_A, hasObjectArrayLabel, insertBeforeInstr);
+
+    insertBeforeInstr->InsertBefore(noObjectArrayLabel);
+}
+
+void
+Lowerer::GenerateInitForInEnumeratorFastPath(IR::Instr * instr, Js::ForInCache * forInCache)
+{
+    Func * func = this->m_func;
+
+    IR::LabelInstr * helperLabel = IR::LabelInstr::New(Js::OpCode::Label, func, true);
+    IR::RegOpnd * objectOpnd = instr->GetSrc1()->AsRegOpnd();
+
+    // Tagged check and object check
+    m_lowererMD.GenerateObjectTest(objectOpnd, instr, helperLabel);
+    m_lowererMD.GenerateIsDynamicObject(objectOpnd, instr, helperLabel);
+
+    // Type check with cache
+    //
+    //   MOV typeOpnd, [objectOpnd + offset(type)]
+    //   CMP [&forInCache->type], typeOpnd
+    //   JNE $helper
+
+    IR::RegOpnd * typeOpnd = IR::RegOpnd::New(TyMachPtr, func);
+    InsertMove(typeOpnd, IR::IndirOpnd::New(objectOpnd, Js::DynamicObject::GetOffsetOfType(), TyMachPtr, func), instr);
+    InsertCompareBranch(IR::MemRefOpnd::New(&forInCache->type, TyMachPtr, func, IR::AddrOpndKindForInCacheType), typeOpnd, Js::OpCode::BrNeq_A, helperLabel, instr);
+    
+    // Check forInCacheData->EnumNonEnumerable == false  
+    //
+    //   MOV forInCacheDataOpnd, [&forInCache->data]
+    //   CMP forInCacheDataOpnd->enumNonEnumerable, 0
+    //   JNE $helper
+
+    IR::RegOpnd * forInCacheDataOpnd = IR::RegOpnd::New(TyMachPtr, func);
+    InsertMove(forInCacheDataOpnd, IR::MemRefOpnd::New(&forInCache->data, TyMachPtr, func, IR::AddrOpndKindForInCacheData), instr);    
+    InsertCompareBranch(IR::IndirOpnd::New(forInCacheDataOpnd, Js::DynamicObjectPropertyEnumerator::GetOffsetOfCachedDataEnumNonEnumerable(), TyUint8, func),
+        IR::IntConstOpnd::New(0, TyUint8, func), Js::OpCode::BrNeq_A, helperLabel, instr);
+        
+    // Check has object array
+    GenerateHasObjectArrayCheck(objectOpnd, typeOpnd, helperLabel, instr);
+
+    // Check first prototype with enumerable properties
+    //
+    //   MOV prototypeObjectOpnd, [type + offset(prototype)]
+    //   MOV prototypeTypeOpnd, [prototypeObjectOpnd + offset(type)]
+    //   CMP [prototypeTypeOpnd + offset(typeId)], TypeIds_Null
+    //   JEQ $noPrototypeWithEnumerablePropertiesLabel
```

Comment:
Got it.

---
In reply to: [81578312](https://github.com/Microsoft/ChakraCore/pull/1641#discussion_r81578312) [](ancestors = 81578312)